src/sudoku.rs

```diff
@@ -100,27 +120,54 @@ impl<const N: usize> Board<N> {
     #[inline]
     #[must_use]
     pub fn fitness(&self) -> u8 {
-        self.count_duplicates() + self.transpose().count_duplicates()
+        self.count_row_duplicates()
+            + self.transpose().count_row_duplicates()
+            + self.count_box_duplicates()
     }
 
     #[inline]
     #[must_use]
-    pub fn count_duplicates(&self) -> u8 {
+    pub fn count_row_duplicates(&self) -> u8 {
         let mut total_duplicates: u8 = 0;
 
         for row in self.0 {
-            let mut duplicates: u8 = 0;
-            let mut duplicate_counts: [u8; N] = [0; N];
+            let mut scorer = Scorer::default();
 
             for value in &row.0 {
-                duplicate_counts[(*value - 1) as usize] += 1;
+                scorer.check(*value);
+            }
+            total_duplicates += scorer.score();
+        }
+
+        total_duplicates
+    }
+
+    fn count_box_duplicates(&self) -> u8 {
+        let mut total_duplicates: u8 = 0;
 
-                if duplicate_counts[(*value - 1) as usize] > 1 {
-                    duplicates += 1;
+        // XXX This could be a proper integer square root.
+        // Realistically these are the only sizes that
+        // matter anyhow, and theres no built-in integer
+        // sqrt() in Rust.
+        let box_size = match N {
+            4 => 2,
+            9 => 3,
+            16 => 4,
+            25 => 5,
+            _ => panic!("puzzle size N must be one of (2..5)**2"),
+        };
+
+        for row in (0..N).step_by(box_size) {
+            for col in (0..N).step_by(box_size) {
+                let mut scorer = Scorer::default();
+
+                for r in &self.0[row..row + box_size] {
```

Comment:
I also like this because it looks like it `+ box_size` is applied to each index in the range. I wouldn't have expected that behavior, but it's cool to see that it works :+1: